### DISCUSSION TITLE
How to output beam forces / moments from ComputeBeamResultants

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#26138] [Q&A Modules: Solid mechanics]
Hello. Tell me please how i can output Forces and Moments in beam element from ComputeBeamResultants for the following visualization in paraview.

**Comment by GiudGiud** [comment]
@cbolisetti @bwspenc

**Reply by andrey1887** [reply]
Thank you @GiudGiud.
@cbolisetti, @bwspenc. Maybe this is possible by specifying some option in the input file?

**Comment by cbolisetti** [comment]
Beam forces and moments are stored as materialproperties and you can output them using Auxkernels. There should be examples if you go through the test files and documentation.

**Reply by andrey1887** [reply]
Thank you @cbolisetti.

**Reply by andrey1887** [reply]
@cbolisetti, can you give a link for some suitable example. Do i need to create auxkernel in my app to do it? What type in auxkernel i must specify?
This is constructor of ComputeBeamResultants:
```
ComputeBeamResultants::ComputeBeamResultants(const InputParameters & parameters)
  : Material(parameters),
    _disp_strain_increment(
        getMaterialPropertyByName<RealVectorValue>("mech_disp_strain_increment")),
    _rot_strain_increment(getMaterialPropertyByName<RealVectorValue>("mech_rot_strain_increment")),
    _material_stiffness(getMaterialPropertyByName<RealVectorValue>("material_stiffness")),
    _material_flexure(getMaterialPropertyByName<RealVectorValue>("material_flexure")),
    _total_rotation(getMaterialPropertyByName<RankTwoTensor>("total_rotation")),
    _force(declareProperty<RealVectorValue>("forces")),
    _moment(declareProperty<RealVectorValue>("moments")),
    _force_old(getMaterialPropertyOld<RealVectorValue>("forces")),
    _moment_old(getMaterialPropertyOld<RealVectorValue>("moments"))
{
}
```

**Reply by andrey1887** [reply]
Now i added the next blocks in input file:
```
[AuxVariables]
  [forces]
    family = MONOMIAL
    order = CONSTANT
  []

[AuxKernels]
  [forces]
    type = MaterialStdVectorAux
    property = forces
    component = 0
    variable = forces
  []
```
And the error occured:
```
*** ERROR ***                                                                                                                         
The following error occurred in the object "forces", of type "MaterialStdVectorAux".                                                  
                                                                                                                                      
The requested non-AD material property 'forces' of type 'std::vector<double>'                                                         
is already retrieved or declared as a non-AD property of type 'libMesh::VectorValue<double>'. 
```

**Reply by GiudGiud** [reply]
Is there ADMaterialStdVectorAux you can use? 
The problem here is that the material property is AD but the auxkernel is retrieving a non-AD

**Reply by andrey1887** [reply]
Thank you, @GiudGiud for your answer, but with AD almost the same thing:
```
*** ERROR ***                                                                                                                         
The following error occurred in the object "forces", of type "ADMaterialStdVectorAux".                                                
                                                                                                                                      
The requested AD material property 'forces' of type 'std::vector<double>'                                                             
is already retrieved or declared as a non-AD property of type 'libMesh::VectorValue<double>'.
```

**Reply by GiudGiud** [reply]
Oh this is an array vs vector thing

Array is stdVector in the title. Remove std from the name of the auxkernel

**Reply by andrey1887** [reply]
```
[AuxKernels]
  [forces]
    type = MaterialRealVectorValueAux
    property = forces
    component = 0
    variable = forces
  []
```
Working variant

**Reply by GiudGiud** [reply]
this one
https://mooseframework.inl.gov/source/auxkernels/MaterialRealVectorValueAux.html

**Reply by GiudGiud** [reply]
or maybe the AD version.
Not sure how we can fix this. Maybe a table of the matprop type vs name of the object ?

**Reply by andrey1887** [reply]
![изображение](https://github.com/idaholab/moose/assets/114737834/7125884d-5703-4846-ae7e-81875943c01b)
It looks normal. But here nodal moments obtained from nearby elements and at the end elements
midpoint values are simply translated to edge nodes. It would be better if the nodal values of moments or forces will be passed into exodus.
![изображение](https://github.com/idaholab/moose/assets/114737834/37ce6fab-2bc1-40d8-bab1-659a61ea4926)

**Reply by GiudGiud** [reply]
Are you outputting hermite variables to exodus?
Not every family is supported at the same level for exodus output, we ll ask Roy after

**Reply by andrey1887** [reply]
No, it's currently Lagrange's ones.

**Reply by cbolisetti** [reply]
We see small discrepancies like that too. Because moments and forces are accurate at the midpoint of the element not at the nodes. So if the elements are small enough, you can get a better match. 

Also, we use `MaterialRealVectorVectorValueAux` for getting the forces and moments out, like @GiudGiud suggested. I don't know if there is a better solution with the current implementation.

**Reply by andrey1887** [reply]
Thank you, @cbolisetti. But, considering following https://github.com/idaholab/moose/discussions/26145#discussioncomment-7669061, is there a possibility to output moments or forces at nodes?

**Reply by cbolisetti** [reply]
What about @hugary1995 suggestion in a different thread about using tagged vectors?

**Reply by andrey1887** [reply]
it is about element's nodal reactions.

**Reply by cbolisetti** [reply]
Wouldn't the forces and moments for an element at the node be the same as nodal reactions for that element?

**Reply by andrey1887** [reply]
Yes, it is the same. So i would like to output it directly from nodes without retrieving from elements by paraview tools.

**Reply by andrey1887** [reply]
Hello.
For example, how these diagrams looks in some FEA software:
<img src="https://github.com/idaholab/moose/assets/114737834/7ed6f556-6845-4ce6-9e59-7d0070d62b25" width="400">
At 25-th node, an external transverse force applied. So there is a gap on the diagram at this node. (In our case it would be slanted line, wich connects midpoint values) 
For each element there are values of shear force at both nodes, which allow to do such thing:
![image](https://github.com/idaholab/moose/assets/114737834/521beaec-e041-4e41-a8cf-0dda3ed01856)

**Reply by GiudGiud** [reply]
So to summarize the problem.

- material properties computed at Qp correctly
- projected onto a nodal lagrange variable using a MaterialRealVectorValueAux
- output is not correct on the boundary nodes. But correct inside the domain

I think the step where there is an issue is the projection. Instead of a projection, we want to be computing the nodal values directly.
We actually do not have nodal material properties because at a node you cannot know which subdomain the material property definition is coming from. 
BUT one can still re-code the definition of the material property in an auxkernel that is being executed at nodes

**Reply by andrey1887** [reply]
Hello. @GiudGiud, i just have seen in `ComputeBeamResultants`, that forces and moments are in global coordinate system: 
```
  /// Current force vector in global coordinate system
  MaterialProperty<RealVectorValue> & _force;

  /// Current moment vector in global coordinate system
  MaterialProperty<RealVectorValue> & _moment;
  ```
  If they are not related to local axes of elements how these diagram how can they be read?
  For example if elements are perpendicular to each other .

**Reply by GiudGiud** [reply]
@cbolisetti 
I would think if they are in the global coordinate system that actually makes the output easier, you just need to also output the local coordinates to interpret them?

**Reply by andrey1887** [reply]
Thank you, @GiudGiud  for your answer. I planned to use local axes from here:  
https://github.com/idaholab/moose/discussions/25653#discussion-5696825
to do it.

**Reply by andrey1887** [reply]
Hello. @GiudGiud, @cbolisetti. Now i try to understand how nodal forces and moments are obtained in `StressDivergenceBeam::computeGlobalResidual`. So in this piece of code:
```
RealVectorValue a;
  _force_local_t.resize(_qrule->n_points());
  _moment_local_t.resize(_qrule->n_points());
  _local_force_res.resize(_test.size());
  _local_moment_res.resize(_test.size());

  // convert forces/moments from global coordinate system to current beam local configuration
  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
  {
    _force_local_t[_qp] = (*total_rotation)[0] * (*force)[_qp];
    _moment_local_t[_qp] = (*total_rotation)[0] * (*moment)[_qp];
  }

  // residual for displacement variables
  for (_i = 0; _i < _test.size(); ++_i)
  {
    _local_force_res[_i] = a;
    for (unsigned int component = 0; component < 3; ++component)
    {
      for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
        _local_force_res[_i](component) +=
            (_i == 0 ? -1 : 1) * _force_local_t[_qp](component) * 0.5;
    }
  }
  ```
  `_local_force_res[_i]` is equating to `a` (which has zero elements) and then a half value of `_force_local_t[_qp](component)` is assigned with minus to 0 node, and another half to 1 node.

<img src="https://github.com/idaholab/moose/assets/114737834/4c5096ae-0418-4be4-8393-67c29bdf9608" width="400">

As i understand `_force_local_t[_qp](component)` comes from `ComputeBeamResultants` and it value determined at the midpoint of element.
May be i am wrong somewhere?
But how the  `_force_local_t[_qp](component)` can be imagined / interpreted?

**Reply by GiudGiud** [reply]
Hello

```
    _local_force_res[_i] = a;
```
seems weird to me. Seems like it s just setting the value to a 0 vector, which we did not really need to pre-allocate and call a vector `a` to do that.

Anyhow,

The `_force_local_t` is coming from the `forces` material property, just rotated into the local coordinates (second paragraph of code you pasted). I imagine you checked that this force was computed in ComputeBeamResultants. 

It's indexed at the _qp. So unless computeBeamResultant is a bit hacky, this should be the value at the quadrature point. What is the quadrature here? Is it special and points are at midpoints?

**Reply by cbolisetti** [reply]
@GiudGiud I think it is a little bit hacky in the context of forces and moments and @andrey1887 is correct. @andrey1887 can you output the global forces and moments and convert them to local yourself? Otherwise you'll have to make _force_local_t into `const MaterialProperty<RealVectorValue> * _force_local_t` and use the auxkernels to output it just like the global values.

**Reply by andrey1887** [reply]
Thank you, @GiudGiud.
There are 2 q-points. FIRST order LAGRANGE variables are used.
Sorry, my assumption about 1 midpoint was wrong. But the force values at both q-points are the same.

**Reply by GiudGiud** [reply]
Could it be that the midpoint value is being set to both Qp values then ? In light of what Chandu said

**Reply by andrey1887** [reply]
Thank you @cbolisetti. Now the forces output is more clear. Would you tell me please, how local force residuals are computed here:
```
  // residual for displacement variables
  for (_i = 0; _i < _test.size(); ++_i)
  {
    _local_force_res[_i] = a;
    for (unsigned int component = 0; component < 3; ++component)
    {
      for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
        _local_force_res[_i](component) +=
            (_i == 0 ? -1 : 1) * _force_local_t[_qp](component) * 0.5;
    }
  }
  ```
  Now i can imagine only nodal reactions of element :
<img src="https://github.com/idaholab/moose/assets/114737834/ff0619e0-d1cf-4509-a6e3-fd43add68646" width="400">

  but not how it computed from _force_local_t[_qp](component)

**Reply by andrey1887** [reply]
@GiudGiud. Simple the single values of gradients in are used for all q-points in 
`ComputeIncrementalBeamStrain::computeQpStrain()`

 ```
 const RealVectorValue grad_rot_0(1.0 / _original_length[0] * (_rot1 - _rot0));
  const RealVectorValue avg_rot(
      0.5 * (_rot0(0) + _rot1(0)), 0.5 * (_rot0(1) + _rot1(1)), 0.5 * (_rot0(2) + _rot1(2)));
  const RealVectorValue grad_disp_0(1.0 / _original_length[0] * (_disp1 - _disp0));

  _grad_disp_0_local_t = _total_rotation[0] * grad_disp_0;
  _grad_rot_0_local_t = _total_rotation[0] * grad_rot_0;
  _avg_rot_local_t = _total_rotation[0] * avg_rot;
  ```
  Thank you @GiudGiud  for reminding me about 2 q-point.

**Reply by GiudGiud** [reply]
Ok so all the Qp values are the midpoint values. Make sense to me then. 

I think we ought to use a 1-point quadrature then. Do we have an issue (and maybe feature request?) for this @cbolisetti

**Reply by cbolisetti** [reply]
So material and cross-section properties are still evaluated at qps so they are not totally useless. That said, there's definitely room for improvement here. Just no funding.
